protocol/consistencychecks.go

```diff
@@ -0,0 +1,321 @@
+// Implements all consistency check operations done by a CONIKS client
+// on data received from a CONIKS directory.
+// These include data binding proof verification,
+// and non-equivocation checks.
+
+package protocol
+
+import (
+	"bytes"
+
+	"github.com/coniks-sys/coniks-go/crypto/sign"
+	"github.com/coniks-sys/coniks-go/crypto/vrf"
+	m "github.com/coniks-sys/coniks-go/merkletree"
+)
+
+const (
+	invalidProof int = iota
+	proofOfAbsence
+	proofOfInclusion
+)
+
+// ConsistencyChecks stores the current consistency check
+// state of a CONIKS client. This includes the current epoch
+// and SignedTreeRoot value, as well as directory's policies
+// (e.g., wherether the TemporaryBinding extension is being used or not)
+// The client should create a new ConsistencyChecks instance only once,
+// when she registers her binding with a ConiksDirectory.
+// This ConsistencyChecks instance then will be used to verify
+// the returned responses from the ConiksDirectory.
+type ConsistencyChecks struct {
+	SavedSTR *m.SignedTreeRoot
+
+	// extensions settings
+	useTBs bool
+	TBs    map[string]*m.TemporaryBinding
+
+	// signing key
+	signKey sign.PublicKey
+}
+
+// NewCC creates an instance of ConsistencyChecks using
+// the pinning directory's STR at epoch 0 or
+// the consistency state read from a persistent storage.
+func NewCC(savedSTR *m.SignedTreeRoot, useTBs bool, signKey sign.PublicKey) *ConsistencyChecks {
+
+	// TODO: see #110
+	if !useTBs {
+		panic("Currently the server is forced to use TBs")
+	}
+
+	cc := &ConsistencyChecks{
+		SavedSTR: savedSTR,
+		useTBs:   useTBs,
+		signKey:  signKey,
+	}
+
+	if useTBs {
+		cc.TBs = make(map[string]*m.TemporaryBinding)
+	}
+	return cc
+}
+
+// UpdateConsistency verifies the directory's response for a request and
+// updates the current state if all the verifications are passed.
+// It first verifies the directory's returned status code of the request.
+// If the status code is not in the ErrorResponses array, it means
+// the directory has successfully handled the request.
+// The verifier will then verify the consistency state of the response.
+// This will panic if it is called with an int
+// which isn't a valid/known request-type.
+func (cc *ConsistencyChecks) UpdateConsistency(requestType int, msg *Response,
+	uname string, key []byte) ErrorCode {
+	if ErrorResponses[msg.Error] {
+		return msg.Error
+	}
+
+	var str *m.SignedTreeRoot
+	var err error
+
+	switch requestType {
+	case RegistrationType:
+		str, err = cc.verifyRegistration(requestType, msg, uname, key)
+	case KeyLookupType:
+		str, err = cc.verifyKeyLookup(requestType, msg, uname, key)
+	case MonitoringType:
+	case KeyLookupInEpochType:
+	default:
+		panic("[coniks] Unknown request type")
+	}
+
+	// update the state
+	if err == PassedWithAProofOfAbsence ||
+		err == PassedWithAProofOfInclusion {
+		cc.SavedSTR = str
```

Comment:
@arlolra 
> we should change things so that it's unnecessary

I'm not sure if it is unnecessary even we do the above change. When the client receives another STR for the current epoch, does it still need to verify whether the STR is the same as the saved STR? If not, the server can perform an equivocation by changing the tree hash in that STR (since the client verifies only its signature).